lambdas/ec2/schema_enforcer.py

```diff
@@ -0,0 +1,152 @@
+from __future__ import print_function
+
+import json
+import boto3
+import datetime
+import time
+import dateutil
+import re
+
+
+def check_for_tag(ec2_instance, tag_name):
+    """
+    :param ec2_instance: a boto3 resource representing an Amazon EC2 Instance.
+    :param tag_name: A string of the key name you are searching for.
+
+    This method returns False if the ec2 instance currently has no tags
+    or if the tag is not found. If the tag is found, it returns the tag
+    value.
+    """
+    ec2_instance.load()
+    if ec2_instance.tags is None:
+        return False
+    for tag in ec2_instance.tags:
+        if tag['Key'] == tag_name:
+            return tag['Value']
+        else:
+            continue
+    return False
+
+def timenow_with_utc():
+    """
+    Return a datetime object that includes the tzinfo for utc time.
+    """
+    time = datetime.datetime.utcnow()
+    time = time.replace(tzinfo=dateutil.tz.tz.tzutc())
+    return time
+
+def wait_for_tags(ec2_instance, wait_time):
+    """
+    :param ec2_instance: a boto3 resource representing an Amazon EC2 Instance
+    :param wait_time: The number of minutes to wait for the 'termination_date'
+
+    This method exits when either a termination date is found or the wait_time
+    has passed. The method looks for the 'lifetime' key, parses it, and sets
+    the 'termination_date' on the instance.
+    """
+    start = timenow_with_utc()
+    timeout = start + datetime.timedelta(minutes=wait_time)
+
+    while timenow_with_utc() < timeout:
+        if type(check_for_tag(ec2_instance, 'termination_date')) is str:
+            print("'termination_date' tag found!")
+            break
+        if check_for_tag(ec2_instance, 'lifetime') is False:
+            print("No 'lifetime' tag found; sleeping for 15s")
+            time.sleep(15)
+            continue
+        print('lifetime tag found')
+        lifetime = check_for_tag(ec2_instance,'lifetime')
+        lifetime_delta = calculate_lifetime_delta(lifetime)
+        termination_date = start + lifetime_delta
+        ec2_instance.create_tags(
+                Tags=[
+                    {
+                        'Key': 'termination_date',
+                        'Value': termination_date.isoformat()
+                    }
+                ]
+            )
+    else:
+        print('No termination_date found; terminating instance')
+        # TODO: This code should also eventually report terminated instances
+        # to some log somewhere.
+        # ec2_instance.terminate()
+        raise Exception('EC2 instance deleted because no termination_date found within 4 minutes of creation.')
```

Comment:
Yeah, the Enforcer deletes it if it doesn't comply with the rules. I think this confusion points to the fact that I should have a Rules section as well, to really spell it out.